### PR TITLE
wallet: Abandon descendants of orphaned coinbases

### LINKED ISSUE #14148: abandontransaction needed after spending orphaned block reward
When a block reward is spent together with other outputs in the wallet and that block reward is later orphaned, the wallet does not show the other funds as available anymore.  Even a `-rescan` is not enough - only `-zapwallettxes` or an explicit `abandontransaction` "recovers" the funds.  I prepared a [test case](https://github.com/domob1812/bitcoin/blob/wallet-bug-invalidated-block-rewards/test/functional/wallet_bug.py) that shows this behaviour.

Is this a known issue or the expected behaviour?  While this situation is something that will presumably not occur in practice (as it requires a long reorg), I wonder if that transaction should be automatically abandoned as it seems clear it won't be able to be confirmed anymore.

### PR DESCRIPTION
**Pull request by achow101**
When a block is reorged out of the main chain, any descendants of the coinbase will no longer be valid. Currently they are only marked as inactive, which means that our balance calculations will still include them. In order to be excluded from the balance calculation, they need to either be abandoned or conflicted. This PR goes with the abandoned method.

Note that even when they are included in balance calculations, coin selection will not select outputs belonging to these transactions because they are not in the mempool.

Fixes #14148

### CODE REVIEW

**Comment by luke-jr** [comment]
Approach NACK I think. Shouldn't these already be conflicted?

**Comment by achow101** [comment]
> Shouldn't these already be conflicted?

Nope, see the linked issue. Anything that is invalidated by a reorg just gets marked as "inactive" which means "unconfirmed and not in mempool".

I looked at marking such txs as conflicted, but ran into some issues with setting the conflicting blockhash and height as `invalidateblock` just marks something as invalid (and reorgs out everything) without having a conflicting block.

**Comment by luke-jr** [comment]
I guess in the context of `invalidateblock`, the transaction should be entirely invalid and thus not in the wallet at all. But that would be ugly, especially after it's already there.

I'm not sure unusual cases like that should get in the way of doing the right thing for the normal conflict case, though. And even that unusual case should resolve cleanly if we handle the normal case, after a competing block is accepting.

**Review by shaavan — COMMENTED** [review]
@achow101 

In the second commit, you mentioned the unconfirmed ancestors should be marked as abandoned even when the orphan coinbase has been reorged into the main chain.

What's the reasoning behind this?

If, say, I mine a block and create descendant transactions from it, and say it got orphaned, so I further mine over it and say I was successful in making it part of the main chain, why should my transactions, created using this block's coinbase transaction should get abandoned?

File: test/functional/wallet_orphanedreward.py [hunk 1]
```diff
@@ -32,28 +32,31 @@ def run_test(self):
...
-          "untrusted_pending": 0,
-          "immature": 0,
-        })
-        # The following abandontransaction is necessary to make the later
-        # lines succeed, and probably should not be needed; see
-        # https://github.com/bitcoin/bitcoin/issues/14148.
-        self.nodes[1].abandontransaction(txid)
+        blocks = self.generate(self.nodes[0], 152)
+        conflict_block = blocks[0]
+        # We expect the desendants of orphaned rewards to no longer be considered
```
Comment (rajarshimaitra) [inline test/functional/wallet_orphanedreward.py:42]: nit:
```suggestion
        # We expect the descendants of orphaned rewards to no longer be considered
```

File: test/functional/wallet_orphanedreward.py [hunk 2]
```diff
@@ -32,28 +32,31 @@ def run_test(self):
...
           "trusted": 10,
           "untrusted_pending": 0,
           "immature": 0,
         })
-        self.nodes[1].sendtoaddress(self.nodes[0].getnewaddress(), 9)
+        # And the unconfirmed tx to be abandoned
+        assert_equal(self.nodes[1].gettransaction(txid)["details"][0]["abandoned"], True)
+
+        # If the orphaned reward is reorged back into the main chain, any unconfirmed txs
+        # at the time of the original reorg remain abandoned.
```
Comment (rajarshimaitra) [inline test/functional/wallet_orphanedreward.py:52]: Should it be more clear to say "any unconfirmed descendant transactions remains abandoned"?

File: test/functional/wallet_orphanedreward.py [hunk 3]
```diff
@@ -32,28 +32,31 @@ def run_test(self):
         self.generate(self.nodes[0], 150)
         assert_equal(self.nodes[1].getbalance(), 10 + 25)
```
Comment (rajarshimaitra) [inline test/functional/wallet_orphanedreward.py:33]: Maybe I am dumb and missing something very obvious, why is the block reward 25 here?

File: test/functional/wallet_orphanedreward.py [hunk 4]
```diff
@@ -32,28 +32,31 @@ def run_test(self):
...
-        self.nodes[1].sendtoaddress(self.nodes[0].getnewaddress(), 9)
+        # And the unconfirmed tx to be abandoned
+        assert_equal(self.nodes[1].gettransaction(txid)["details"][0]["abandoned"], True)
+
+        # If the orphaned reward is reorged back into the main chain, any unconfirmed txs
+        # at the time of the original reorg remain abandoned.
+        self.nodes[0].invalidateblock(conflict_block)
+        self.nodes[0].reconsiderblock(blk)
+        assert_equal(self.nodes[0].getbestblockhash(), orig_chain_tip)
+        self.generate(self.nodes[0], 3)
```
Comment (rajarshimaitra) [inline test/functional/wallet_orphanedreward.py:56]: Should we let the block reward mature again?? Then show that the reward is included in the balance but the descendant is still abandoned?  I manually verified that's the case..

**Review by rajarshimaitra — COMMENTED** [review]
Concept + tACK b5e49a4b31debdc1801a22cc1de1542ab5f205c6

Few nits, questions and suggestions below.. 

> Leaving the transactions abandoned just means that the user will have to rebroadcast the transactions themselves. Considering the rarity of this situation in practice, I don't think it's worth the effort to implement anything more complex either.

Should this be documented somewhere? User maybe actively noticing the situation as you suggested but might not know that the descendant needs another rebroadcast? Also given its a 100 blk reorg, the possibility of this occurring is very low.. But maybe just for sake completeness, makes sense to write a line in the code somewhere explaining the manual broadcasting part?

File: test/functional/wallet_orphanedreward.py (hunk 3 above)
Comment (achow101) [reply test/functional/wallet_orphanedreward.py:33]: Regtest's halving interval is 150 blocks.

File: test/functional/wallet_orphanedreward.py (hunk 1 above)
Comment (achow101) [reply test/functional/wallet_orphanedreward.py:42]: Done

**Comment by achow101** [comment]
> Should this be documented somewhere? User maybe actively noticing the situation as you suggested but might not know that the descendant needs another rebroadcast? Also given its a 100 blk reorg, the possibility of this occurring is very low.. But maybe just for sake completeness, makes sense to write a line in the code somewhere explaining the manual broadcasting part?

I've added a comment about that, as well as a comment that the transaction states can still change after abandoning.

**Comment by MarcoFalke** [comment]
CI:

```
  File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/wallet_orphanedreward.py", line 57, in run_test
    self.generate(self.nodes[0], 3)
  File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 650, in generate
    sync_fun() if sync_fun else self.sync_all()
  File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 715, in sync_all
    self.sync_mempools(nodes)
  File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 708, in sync_mempools
    raise AssertionError("Mempool sync timed out after {}s:{}".format(
AssertionError: Mempool sync timed out after 2400s:
  {'22b1bf7d8f18b9a08c8f5ad676f3372000a5ad247f4fae848b2a76309c0dba3b'}
  set()

File: test/functional/wallet_orphanedreward.py (hunk 3 above)
Comment (rajarshimaitra) [reply test/functional/wallet_orphanedreward.py:33]: Ohh.. My bad..

**Comment by achow101** [comment]
> CI:
> 
> ```
>   File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/wallet_orphanedreward.py", line 57, in run_test
>     self.generate(self.nodes[0], 3)
>   File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 650, in generate
>     sync_fun() if sync_fun else self.sync_all()
>   File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 715, in sync_all
>     self.sync_mempools(nodes)
>   File "/tmp/cirrus-ci-build/ci/scratch/build/bitcoin-x86_64-pc-linux-gnu/test/functional/test_framework/test_framework.py", line 708, in sync_mempools
>     raise AssertionError("Mempool sync timed out after {}s:{}".format(
> AssertionError: Mempool sync timed out after 2400s:
>   {'22b1bf7d8f18b9a08c8f5ad676f3372000a5ad247f4fae848b2a76309c0dba3b'}
>   set()
> ```

This error is actually really confusing. I ran into it a few times while writing the test, but whenever I actually tried to debug it, I couldn't get it to reproduce.

What's happening is that the unconfirmed transaction descended from the previously orphaned parent is somehow making its way back into node0's mempool. What I don't understand is how that is happening. When attempting to debug this, I added some `print(self.nodes[0].getrawmempool())` statements before the generate and those indicated that node0's mempool is empty, as it should be. The transaction is removed from the mempool when its parent is reorged out of the chain, and the wallet is not rebroadcasting it either, so I don't understand how node0 becomes aware of it again.

**Review by rajarshimaitra — APPROVED** [review]
ReACK 1a8f1148f00af8727ab5e2d58ae71bc69a75fdae

**Review by aureleoules — APPROVED** [review]
ACK 1a8f1148f00af8727ab5e2d58ae71bc69a75fdae

I verified that the test correctly test the behavior. Also executed the test hundreds of times in parallel in case there are intermittent failures and couldn't find one. (they are usually rare to reproduce though)

---

I suggest this diff (uses auto, structured binding and avoids some indirections)

<details>
<summary>Details</summary>

```diff
diff --git a/src/wallet/wallet.cpp b/src/wallet/wallet.cpp
index 8c5c7e445..a5787b3a7 100644
--- a/src/wallet/wallet.cpp
+++ b/src/wallet/wallet.cpp
@@ -1072,7 +1072,7 @@ CWalletTx* CWallet::AddToWallet(CTransactionRef tx, const TxState& state, const
         std::vector<CWalletTx*> txs;
         txs.push_back(&wtx);
 
-        TxStateInactive inactive_state = TxStateInactive{/*abandoned=*/true};
+        const auto inactive_state = TxStateInactive{/*abandoned=*/true};
 
         while (!txs.empty()) {
             CWalletTx* tx = txs.back();
@@ -1081,10 +1081,11 @@ CWalletTx* CWallet::AddToWallet(CTransactionRef tx, const TxState& state, const
             // Break caches since we have changed the state
             tx->MarkDirty();
             MarkInputsDirty(tx->tx);
+            const auto& tx_hash = tx->GetHash();
             for (unsigned int i = 0; i < tx->tx->vout.size(); ++i) {
-                COutPoint outpoint(tx->GetHash(), i);
-                std::pair<TxSpends::const_iterator, TxSpends::const_iterator> range = mapTxSpends.equal_range(outpoint);
-                for (TxSpends::const_iterator it = range.first; it != range.second; ++it) {
+                const COutPoint outpoint(tx_hash, i);
+                const auto& [begin, end] = mapTxSpends.equal_range(outpoint);
+                for (auto it = begin; it != end; ++it) {
                     const auto wit = mapWallet.find(it->second);
                     if (wit != mapWallet.end()) {
                         txs.push_back(&wit->second);
```

</details>

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -1067,6 +1067,33 @@ CWalletTx* CWallet::AddToWallet(CTransactionRef tx, const TxState& state, const
...
 
+    // Mark inactive coinbase transactions and their descendants as abandoned
+    if (wtx.IsCoinBase() && wtx.isInactive()) {
+        std::vector<CWalletTx*> txs;
+        txs.push_back(&wtx);
+
+        TxStateInactive inactive_state = TxStateInactive{/*abandoned=*/true};
+
+        while (!txs.empty()) {
+            CWalletTx* tx = txs.back();
```
Comment (furszy) [inline src/wallet/wallet.cpp:1078]: would be better to rename this variable, it's shadowing the method's argument.

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -1067,6 +1067,33 @@ CWalletTx* CWallet::AddToWallet(CTransactionRef tx, const TxState& state, const
         }
     }
 
+    // Mark inactive coinbase transactions and their descendants as abandoned
+    if (wtx.IsCoinBase() && wtx.isInactive()) {
+        std::vector<CWalletTx*> txs;
+        txs.push_back(&wtx);
```
Comment (furszy) [inline src/wallet/wallet.cpp:1073]: nit: could inline this
```c++
std::vector<CWalletTx*> txs{&wtx};
```

**Review by furszy — COMMENTED** [review]
Seems that the coinbase descendants state modification is not permitted to disk. So after a restart, they will get back to their previous state.

Is that intended because we have a duplicated check somewhere around the wallet load process or just something that is missing?

File: src/wallet/wallet.cpp (hunk 5 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Fixed.

File: src/wallet/wallet.cpp (hunk 6 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Done

**Comment by achow101** [comment]
> Seems that the coinbase descendants state modification is not permitted to disk. So after a restart, they will get back to their previous state.

Oops, fixed.

**Review by furszy — APPROVED** [review]
ACK b0fa5989 with a not-blocking nit.

The initial round of the loop (when we only have the orphan coinbase in the vector) isn't needed. We update the state and reset the cached amounts below.

**Review by aureleoules — APPROVED** [review]
reACK b0fa5989e1b77a343349bd36f8bc407f9366a589

File: test/functional/wallet_orphanedreward.py [hunk 7]
```diff
@@ -31,19 +31,40 @@ def run_test(self):
...
+        self.nodes[1].unloadwallet(self.default_wallet_name)
+        self.nodes[1].loadwallet(self.default_wallet_name)
+        assert_equal(self.nodes[1].gettransaction(txid)["details"][0]["abandoned"], True)
+
+        # If the orphaned reward is reorged back into the main chain, any unconfirmed
+        # descendant txs at the time of the original reorg remain abandoned.
+        self.nodes[0].invalidateblock(conflict_block)
+        self.nodes[0].reconsiderblock(blk)
+        assert_equal(self.nodes[0].getbestblockhash(), orig_chain_tip)
+        self.generate(self.nodes[0], 3)
```
[the remark this answers is not in the document]
Comment (achow101) [reply test/functional/wallet_orphanedreward.py:63]: Yes, 2 would work, I think I just miscounted when writing this originally.

**Comment by achow101** [comment]
> The initial round of the loop (when we only have the orphan coinbase in the vector) isn't needed. We update the state and reset the cached amounts below.

It does, but I think it's easier to read this way and the performance impact of breaking the caches twice and writing it twice is not that big.

File: src/wallet/wallet.cpp [hunk 8]
```diff
@@ -1067,6 +1067,33 @@ CWalletTx* CWallet::AddToWallet(CTransactionRef tx, const TxState& state, const
...
+        TxStateInactive inactive_state = TxStateInactive{/*abandoned=*/true};
+
+        while (!txs.empty()) {
+            CWalletTx* desc_tx = txs.back();
+            txs.pop_back();
+            desc_tx->m_state = inactive_state;
+            // Break caches since we have changed the state
+            desc_tx->MarkDirty();
+            batch.WriteTx(*desc_tx);
+            MarkInputsDirty(desc_tx->tx);
```
Comment (ishaanam) [inline src/wallet/wallet.cpp:1083]: Is this needed?

**Review by ishaanam — COMMENTED** [review]
ACK b0fa5989e1b77a343349bd36f8bc407f9366a589

**Pull request merged**